tests/test_unit_conversion.py

```diff
@@ -1,80 +1,1235 @@
+            state_symbols = local_model.get_state_symbols()
+            assert len(state_symbols) == 1
+            assert symbol_a in state_symbols
+            return True
+
+        mV_unit = local_model.get_units('mV')
+        volt_unit = local_model.get_units('volt')
+        original_var = local_model.get_symbol_by_name('env_ode$sv1')
+
+        assert test_original_state(local_model)
+        # test no change in units
+        newvar = local_model.convert_variable('sv11', mV_unit, DataDirectionFlow.INPUT)
+        assert newvar == original_var
+        assert test_original_state(local_model)
+
+        # change mV to V
+        newvar = local_model.convert_variable('sv11', volt_unit, DataDirectionFlow.INPUT)
+        assert newvar != original_var
+        assert len(local_model.variables()) == 5
+        symbol_a = local_model.get_symbol_by_cmeta_id('sv11')
+        assert local_model.get_initial_value(symbol_a) == 0.002
+        assert symbol_a.units == 'volt'
+        assert symbol_a.name == 'env_ode$sv1_converted'
+        symbol_t = local_model.get_symbol_by_cmeta_id('time')
+        assert symbol_t.units == 'ms'
+        symbol_orig = local_model.get_symbol_by_name('env_ode$sv1')
+        assert symbol_orig.units == 'mV'
+        assert local_model.get_initial_value(symbol_orig) == 2.0
+        symbol_derv = local_model.get_symbol_by_name('env_ode$sv1_orig_deriv')
+        assert symbol_derv.units == 'mV / ms'
+        assert not local_model.get_initial_value(symbol_derv)
+        assert len(local_model.equations) == 3
+        assert str(local_model.equations[0]) == 'Eq(_env_ode$sv1, 1000.0*_env_ode$sv1_converted)'
+        assert str(local_model.equations[1]) == 'Eq(_env_ode$sv1_orig_deriv, _1.0)'
+        assert str(local_model.equations[2]) == 'Eq(Derivative(_env_ode$sv1_converted, _environment$time), ' \
+                                                '0.001*_env_ode$sv1_orig_deriv)'
+
+        state_symbols = local_model.get_state_symbols()
+        assert len(state_symbols) == 1
+        assert symbol_a in state_symbols
+
+    def test_add_input_free_variable(self, local_model):
+        """ Tests the Model.convert_variable function that changes units of given variable.
+        This particular case tests changing a free variable
+        For example::
+
+            Original model
+                var{time} time: ms {pub: in};
+                var{sv11} sv1: mV {init: 2};
+
+                ode(sv1, time) = 1{mV_per_ms};
+
+        convert_variable(time, second, DataDirectionFlow.INPUT)
+
+        becomes
+                var time: ms;
+                var{time} time_converted: s;
+                var{sv11} sv1: mV {init: 2};
+                var sv1_orig_deriv mV_per_ms
+
+                time = 1000 * time_converted;
+                sv1_orig_deriv = 1{mV_per_ms}
+                ode(sv1, time_converted) = 1000 * sv1_orig_deriv
+        """
+        # original state
+        def test_original_state(local_model):
```

Comment:
Since this is essentially the same as with the tests above, perhaps extract as a method - it just needs a name that doesn't start with `test_`, e.g. `_check_original_state`?